---
title: "Introduction"
subtitle: "A brief introduction to the workshop"
pagetitle: "Introduction | Tools for Reproducible Research"
description: ""
date: ""
number-sections: false
format: html
---

Welcome to the tutorials! Here we will learn how to make a computational
research project reproducible using several different tools, described in the
figure below:

![](images/reproducibility_overview_with_logos.png){ width=700px }

The figure above gives an overview of the different parts of computational
reproducibility (data, code, workflow and environment), as well as the various
tools that are used for each part; Git is, arguably, integral to all of the
parts, but we only listed it in the _code_ section for a less cluttered figure.

The course has a tutorial for each of the tools, all made so that they can be
completed independently of each other. It is therefore perfectly
possible to go through them in whatever order you prefer, but we suggest the
following order:

 1. Git
 2. Conda
 3. Containers
 4. Quarto
 5. Jupyter
 6. Snakemake
 7. Nextflow

Please make sure to carefully follow the [pre-course
setup](../home_precourse.html) to install the tools and download the course
material before starting with any of the tutorials. These will create quite a
lot of files on your computer, some of which will actually take up a bit of
storage space too. In order to remove any traces of these after completing the
tutorials, please refer to the [take-down section](take-down.html).

## The case study

We will be running a small bioinformatics project as a case study in some of the
tutorials, and use that to exemplify the different steps of setting up a
reproducible research project. To give you some context, the study background
and analysis steps are briefly described below.

### Background

The data is taken from [Osmundson, Dewell, and Darst (2013)](
http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0076572),
who have studied methicillin-resistant _Staphylococcus aureus_ (MRSA). MRSA is
resistant to broad spectrum beta-lactam antibiotics and lead to
difficult-to-treat infections in humans. Lytic bacteriophages have been
suggested as potential therapeutic agents, or as the source of novel antibiotic
proteins or peptides. One such protein, gp67, was identified as
a transcription-inhibiting transcription factor with an antimicrobial effect.
To identify _S. aureus_ genes repressed by gp67, the authors expressed gp67 in
_S. aureus_ cells. RNA-seq was then performed on three _S. aureus_ strains:

- RN4220 with pRMC2 with gp67
- RN4220 with empty pRMC2
- NCTC8325-4

### Analysis

The graph below shows the different steps of the analysis that are included in
this project:

![](images/mrsa-overview.png){ width=400px }

The input files are:

- RNA-seq raw data (FASTQ files) for the three strains
- _S. aureus_ genome sequence (a FASTA file)
- _S. aureus_ genome annotation (a GFF file)

The workflow itself will perform the following tasks:

- Downloading and indexing of the reference genome using _Bowtie2_
- Downloading the raw FASTQ data from the _Sequence Read Archive_ (SRA)
- Run some quality controls on the data using _FastQC_ and _MultiQC_
- Align the raw data to the genome and calculate the gene expression using
  _featureCounts_
- Produce supplementary materials using data from quality controls, gene
  expression and the workflow figure shown above
